File: app/src/main/java/me/jfenn/alarmio/views/ProgressTextView.java
package me.jfenn.alarmio.views;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.LinearInterpolator;

import com.afollestad.aesthetic.Aesthetic;

import androidx.annotation.Nullable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import me.jfenn.alarmio.interfaces.Subscribblable;
import me.jfenn.androidutils.DimenUtils;

public class ProgressTextView extends View implements Subscribblable {

    private Paint linePaint, circlePaint, referenceCirclePaint, backgroundPaint, textPaint;
    private long progress, maxProgress, referenceProgress;

    private String text;

    private int padding;

    private Disposable colorAccentSubscription;
    private Disposable textColorPrimarySubscription;
    private Disposable textColorSecondarySubscription;

    public ProgressTextView(Context context) {
        this(context, null);
    }

    public ProgressTextView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ProgressTextView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        padding = DimenUtils.dpToPx(4);

        linePaint = new Paint();
        linePaint.setAntiAlias(true);
        linePaint.setStyle(Paint.Style.STROKE);
        linePaint.setStrokeWidth(padding);

        circlePaint = new Paint();
        circlePaint.setAntiAlias(true);

        referenceCirclePaint = new Paint();
        referenceCirclePaint.setAntiAlias(true);

        backgroundPaint = new Paint();
        backgroundPaint.setAntiAlias(true);
        backgroundPaint.setStyle(Paint.Style.STROKE);
        backgroundPaint.setStrokeWidth(padding);

        textPaint = new Paint();
        textPaint.setAntiAlias(true);
        textPaint.setTextAlign(Paint.Align.CENTER);
        textPaint.setTextSize(DimenUtils.spToPx(34));
        textPaint.setFakeBoldText(true);

        subscribe();
    }

    @Override
    public void subscribe() {
        colorAccentSubscription = Aesthetic.Companion.get()
                .colorAccent()
                .subscribe(new Consumer<Integer>() {
                    @Override
                    public void accept(Integer integer) throws Exception {
                        linePaint.setColor(integer);
                        circlePaint.setColor(integer);
                        invalidate();
                    }
                });

        textColorPrimarySubscription = Aesthetic.Companion.get()
                .textColorPrimary()
                .subscribe(new Consumer<Integer>() {
                    @Override
                    public void accept(Integer integer) throws Exception {
                        textPaint.setColor(integer);
                        referenceCirclePaint.setColor(integer);
                        invalidate();
                    }
                });

        textColorSecondarySubscription = Aesthetic.Companion.get()
                .textColorSecondary()
                .subscribe(new Consumer<Integer>() {
                    @Override
                    public void accept(Integer integer) throws Exception {
                        backgroundPaint.setColor(integer);
                        backgroundPaint.setAlpha(50);
                        invalidate();
                    }
                });
    }

    @Override
    public void unsubscribe() {
        colorAccentSubscription.dispose();
        textColorPrimarySubscription.dispose();
        textColorSecondarySubscription.dispose();
    }

    public void setText(String text) {
        this.text = text;
        invalidate();
    }

    public void setProgress(long progress) {
        setProgress(progress, false);
    }

    public void setProgress(long progress, boolean animate) {
        if (animate) {
            ValueAnimator animator = ValueAnimator.ofFloat(this.progress, progress);
            animator.setInterpolator(new LinearInterpolator());
            animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                @Override
                public void onAnimationUpdate(ValueAnimator valueAnimator) {
                    ProgressTextView.this.progress = (long) (float) valueAnimator.getAnimatedValue();
                    invalidate();
                }
            });
            animator.start();
        } else {
            this.progress = progress;
            invalidate();
        }
    }

    public void setMaxProgress(long maxProgress) {
        setMaxProgress(maxProgress, false);
    }

    public void setMaxProgress(long maxProgress, boolean animate) {
        if (animate) {
            ValueAnimator animator = ValueAnimator.ofFloat(this.maxProgress, maxProgress);
            animator.setInterpolator(new LinearInterpolator());
            animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                @Override
                public void onAnimationUpdate(ValueAnimator valueAnimator) {
                    ProgressTextView.this.maxProgress = (long) (float) valueAnimator.getAnimatedValue();
                    invalidate();
                }
            });
            animator.start();
        } else {
            this.maxProgress = maxProgress;
            invalidate();
        }
    }

    public void setReferenceProgress(long referenceProgress) {
        this.referenceProgress = referenceProgress;
        invalidate();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        int size = getMeasuredWidth();
        setMeasuredDimension(size, size);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        int size = Math.min(canvas.getWidth(), canvas.getHeight());
        int sidePadding = padding * 3;
        canvas.drawCircle(size / 2, size / 2, (size / 2) - sidePadding, progress > maxProgress && maxProgress > 0 ? linePaint : backgroundPaint);

        if (maxProgress > 0) {
            float angle = (360f * progress / maxProgress);
            float referenceAngle = (360f * referenceProgress / maxProgress);

            Path path = new Path();
            path.arcTo(new RectF(sidePadding, sidePadding, size - sidePadding, size - sidePadding), -90, angle, true);
            canvas.drawPath(path, linePaint);

            canvas.drawCircle((size / 2) + ((float) Math.cos((angle - 90) * Math.PI / 180) * ((size / 2) - sidePadding)), (size / 2) + ((float) Math.sin((angle - 90) * Math.PI / 180) * ((size / 2) - sidePadding)), 2 * padding, circlePaint);
            if (referenceProgress != 0)
                canvas.drawCircle((size / 2) + ((float) Math.cos((referenceAngle - 90) * Math.PI / 180) * ((size / 2) - sidePadding)), (size / 2) + ((float) Math.sin((referenceAngle - 90) * Math.PI / 180) * ((size / 2) - sidePadding)), 2 * padding, referenceCirclePaint);
        }

        if (text != null)
            canvas.drawText(text, size / 2, (size / 2) - ((textPaint.descent() + textPaint.ascent()) / 2), textPaint);
    }

}

File: app/src/main/java/me/jfenn/alarmio/views/ProgressTextView.kt
package me.jfenn.alarmio.views

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import android.view.animation.LinearInterpolator

import com.afollestad.aesthetic.Aesthetic
import io.reactivex.disposables.Disposable
import me.jfenn.alarmio.interfaces.Subscribblable
import me.jfenn.androidutils.DimenUtils

/**
 * Display a progress circle with text in
 * the center.
 */
class ProgressTextView : View, Subscribblable {

    private var progress: Long = 0
    private var maxProgress: Long = 0
    private var referenceProgress: Long = 0
    private var text: String? = null
    private var padding: Int = DimenUtils.dpToPx(4f)

    private var linePaint: Paint = Paint().apply {
        isAntiAlias = true
        style = Paint.Style.STROKE
        strokeWidth = padding.toFloat()
    }

    private var circlePaint: Paint = Paint().apply {
        isAntiAlias = true
    }

    private var referenceCirclePaint: Paint = Paint().apply {
        isAntiAlias = true
    }

    private var backgroundPaint: Paint = Paint().apply {
        isAntiAlias = true
        style = Paint.Style.STROKE
        strokeWidth = padding.toFloat()
    }

    private var textPaint: Paint = Paint().apply {
        isAntiAlias = true
        textAlign = Paint.Align.CENTER
        textSize = DimenUtils.spToPx(34f).toFloat()
        isFakeBoldText = true
    }

    private var colorAccentSubscription: Disposable? = null
    private var textColorPrimarySubscription: Disposable? = null
    private var textColorSecondarySubscription: Disposable? = null

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun subscribe() {
        colorAccentSubscription = Aesthetic.get()
                .colorAccent()
                .subscribe { integer ->
                    linePaint.color = integer
                    circlePaint.color = integer
                    invalidate()
                }

        textColorPrimarySubscription = Aesthetic.get()
                .textColorPrimary()
                .subscribe { integer ->
                    textPaint.color = integer
                    referenceCirclePaint.color = integer
                    invalidate()
                }

        textColorSecondarySubscription = Aesthetic.get()
                .textColorSecondary()
                .subscribe { integer ->
                    backgroundPaint.color = integer
                    backgroundPaint.alpha = 50
                    invalidate()
                }
    }

    override fun unsubscribe() {
        colorAccentSubscription?.dispose()
        textColorPrimarySubscription?.dispose()
        textColorSecondarySubscription?.dispose()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        subscribe()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        unsubscribe()
    }

    /**
     * Set the text (time) to display in the center
     * of the view.
     */
    fun setText(text: String) {
        this.text = text
        invalidate()
    }

    /**
     * Set the current progress value.
     */
    @JvmOverloads
    fun setProgress(progress: Long, animate: Boolean = false) {
        if (animate) {
            ValueAnimator.ofFloat(this.progress.toFloat(), progress.toFloat()).apply {
                interpolator = LinearInterpolator()
                addUpdateListener { valueAnimator ->
                    (valueAnimator.animatedValue as? Float)?.toLong()?.let { value ->
                        setProgress(value, false)
                    }
                }
                start()
            }
        } else {
            this.progress = progress
            postInvalidate()
        }
    }

    /**
     * Set the largest progress that has been acquired so far.
     */
    @JvmOverloads
    fun setMaxProgress(maxProgress: Long, animate: Boolean = false) {
        if (animate) {
            ValueAnimator.ofFloat(this.maxProgress.toFloat(), maxProgress.toFloat()).apply {
                interpolator = LinearInterpolator()
                addUpdateListener { valueAnimator ->
                    (valueAnimator.animatedValue as? Float)?.toLong()?.let { value ->
                        setMaxProgress(value, false)
                    }
                }
                start()
            }
        } else {
            this.maxProgress = maxProgress
            postInvalidate()
        }
    }

    /**
     * Set the progress value of the reference dot (?) on
     * the circle. Mostly used in the stopwatch, to indicate
     * the previous/best lap time.
     */
    fun setReferenceProgress(referenceProgress: Long) {
        this.referenceProgress = referenceProgress
        postInvalidate()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

        val size = measuredWidth
        setMeasuredDimension(size, size)
    }

    override fun onDraw(canvas: Canvas) {
        val size = Math.min(canvas.width, canvas.height)
        val sidePadding = padding * 3
        canvas.drawCircle((size / 2).toFloat(), (size / 2).toFloat(), (size / 2 - sidePadding).toFloat(), if (maxProgress in 1..(progress - 1)) linePaint else backgroundPaint)

        if (maxProgress > 0) {
            val angle = 360f * progress / maxProgress
            val referenceAngle = 360f * referenceProgress / maxProgress

            val path = Path()
            path.arcTo(RectF(sidePadding.toFloat(), sidePadding.toFloat(), (size - sidePadding).toFloat(), (size - sidePadding).toFloat()), -90f, angle, true)
            canvas.drawPath(path, linePaint)

            canvas.drawCircle(size / 2 + Math.cos((angle - 90) * Math.PI / 180).toFloat() * (size / 2 - sidePadding), size / 2 + Math.sin((angle - 90) * Math.PI / 180).toFloat() * (size / 2 - sidePadding), (2 * padding).toFloat(), circlePaint)
            if (referenceProgress != 0L)
                canvas.drawCircle(size / 2 + Math.cos((referenceAngle - 90) * Math.PI / 180).toFloat() * (size / 2 - sidePadding), size / 2 + Math.sin((referenceAngle - 90) * Math.PI / 180).toFloat() * (size / 2 - sidePadding), (2 * padding).toFloat(), referenceCirclePaint)
        }

        text?.let { str ->
            canvas.drawText(str, (size / 2).toFloat(), size / 2 - (textPaint.descent() + textPaint.ascent()) / 2, textPaint)
        }
    }

}
